File: src/MindBloom.Application/Common/Models/PagedResponse.cs
namespace MindBloom.Application.Common.Models;

public sealed class PagedResponse<T>
{
    public List<T> Items { get; init; } = [];

    public int PageNumber { get; init; }

    public int PageSize { get; init; }

    public int TotalCount { get; init; }

    public int TotalPages { get; init; }

    public bool HasPreviousPage =>
        PageNumber > 1;

    public bool HasNextPage =>
        PageNumber < TotalPages;

    public static PagedResponse<T> Create(
        List<T> items,
        int pageNumber,
        int pageSize,
        int totalCount)
    {
        return new PagedResponse<T>
        {
            Items =
                items,

            PageNumber =
                pageNumber,

            PageSize =
                pageSize,

            TotalCount =
                totalCount,

            TotalPages =
                totalCount == 0
                    ? 0
                    : (int)Math.Ceiling(
                        totalCount /
                        (double)pageSize)
        };
    }
}

File: frontend/mindbloom_mobile/lib/core/models/paged_response.dart
class PagedResponse<T> {
  final List<T> items;
  final int pageNumber;
  final int pageSize;
  final int totalCount;
  final int totalPages;
  final bool hasPreviousPage;
  final bool hasNextPage;

  const PagedResponse({
    required this.items,
    required this.pageNumber,
    required this.pageSize,
    required this.totalCount,
    required this.totalPages,
    required this.hasPreviousPage,
    required this.hasNextPage,
  });

  factory PagedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic> json) fromJson,
  ) {
    final rawItems = json['items'];
    final pageNumber = json['pageNumber'] as int? ?? 1;
    final totalPages = json['totalPages'] as int? ?? 0;

    return PagedResponse<T>(
      items: rawItems is List
          ? rawItems
                .whereType<Map>()
                .map((item) => fromJson(Map<String, dynamic>.from(item)))
                .toList()
          : [],
      pageNumber: pageNumber,
      pageSize: json['pageSize'] as int? ?? 10,
      totalCount: json['totalCount'] as int? ?? 0,
      totalPages: totalPages,
      hasPreviousPage: json['hasPreviousPage'] as bool? ?? pageNumber > 1,
      hasNextPage: json['hasNextPage'] as bool? ?? pageNumber < totalPages,
    );
  }
}
